### DISCUSSION TITLE
Calculation of the sum of the numbers inside the numerical nodes and sub-branches

### DISCUSSION THREAD

**Discussion by aaa1386** [freeplane/freeplane#1448] [Help]
Hello, pay attention to the picture below. I want to define a style with yellow color and named Sum that will automatically calculate the sum of all nodes and branches after it that have numbers and are green or blue. But ignore the numbers without blue and green styles

It should be noted that this yellow style, which is called Sum, can be placed anywhere on the mental map, even in several places, but in the end, the sum of green and blue nodes is important to me.

Is it possible to write such a script?
I didn't work with the formulas part of the Freeplan and I don't know, maybe it is possible through it too.
Thanks.

![image](https://github.com/freeplane/freeplane/assets/38586259/834ab088-8874-4121-9c46-50d5f0e4e72b)

**Comment by macmarrum** [comment]
I would use a formula (like below) to do the calculation inside your yellow node and then clone the node as many times as needed (`Ctrl+Shift+C` to copy a single node and `Ctrl+D` to create a clone).

```groovy
=node.find { it.style.backgroundColorCode in ['#66ff33', '#3399ff'] }.collect { it.to.num }.sum(0)
```

**Comment by aaa1386** [comment]
I don't know how to use this formula
Can you explain?

**Reply by macmarrum** [reply]
To learn more about formulas in Freeplane, you can have a look at https://docs.freeplane.org/scripting/Formulas.html and https://docs.freeplane.org/scripting/api-groovy-tutorial.html

To just have it working, copy the formula from the code block above and paste it to the core of your yellow node, instead of the number.

**Comment by aaa1386** [comment]
Very good and interesting.
Thank you.

**Comment by aaa1386** [comment]
I changed the formula to work with the style name.
=node.find { it.style.name  in ['numberaaa' , 'numberbbb'] }.collect { it.to.num }.sum(0)


The only problem is if the nodes whose sum of numbers should be added together, if even one of them is not a number and is a text, the formula will not work and will give an error.
Is there a way to ignore nodes that are text?

**Reply by macmarrum** [reply]
You can use `num0` instead of `num` -> https://docs.freeplane.org/api/org/freeplane/api/Convertible.html#getNum0()

**Comment by aaa1386** [comment]
I have defined the formula to sum two nodes regardless of whether they are style or background.
=node.find { it.id  in ['ID_201987351' , 'ID_7760154'] }.collect { it.to.num0 }.sum(0)
